snekbox/nsjail.py

```diff
@@ -128,13 +128,19 @@ def _consume_stdout(nsjail: subprocess.Popen) -> str:
 
             if output_size > OUTPUT_MAX:
                 # Terminate the NsJail subprocess with SIGKILL.
+                log.info(f"Output exceeded the output limit, sending SIGKILL to NsJail.")
                 nsjail.kill()
                 break
 
         # Ensure that we wait for the NsJail subprocess to terminate.
         nsjail.wait()
 
-        return "".join(output)
+        # When you send signal `N` to a subprocess to terminate it using Popen, it
+        # will return `-N` as its exit code. As we normally get `N + 128` back, we
+        # convert negative exit codes to the `N + 128` form.
+        returncode = -nsjail.returncode + 128 if nsjail.returncode < 0 else nsjail.returncode
```

Comment:
Why here and not in the main function? I'd rather avoid returning a tuple.

Reply:
> I'd rather avoid returning a tuple.

I'll change it, but what is your motivation for disliking it? Just personal preference?

Reply:
I tend to associate it with functions that are broader in scope than they should be. It could also be a sign of clumsiness, like when returning some sort of boolean in addition to data. It may indicate a function can be further refactored. It's also more difficult to document, especially when not using named tuples. Returning two ostensibly unrelated things in a tuple makes the purpose of the function less clear.

In fact, ideally this function would only be concerned with the output and not do anything to the process, but killing the process within doesn't seem avoidable. This also adds to the trouble of clearly documenting it: this function mainly returns the output, but, by the way, it may also kill the process. That may be an important distinction which wouldn't be clear at a glance.